zerver/middleware.py

```diff
@@ -328,29 +331,58 @@ def csrf_failure(request, reason=""):
 class RateLimitMiddleware(MiddlewareMixin):
     def process_response(self, request, response):
         # type: (HttpRequest, HttpResponse) -> HttpResponse
-        if not settings.RATE_LIMITING:
+        if hasattr(request, 'user'):
+            entity = RateLimitedUser(request.user)
+            response = self._process_response(entity, request, response)
+
+        if settings.IP_RATE_LIMITING:
+            ip = get_ip(request)
+            if ip:
+                response  = self._process_response(
+                    RateLimitedIP(ip), request, response, prefix='_ip')
+
+        if settings.EMAIL_RATE_LIMITING:
+            prefix = '_email'
+            attr = '{}_ratelimit_value'.format(prefix)
+            if hasattr(request, attr):
+                email = getattr(request, attr)
+                response  = self._process_response(
+                    RateLimitedEmail(email), request, response, prefix=prefix)
+
+        return response
+
+    def _process_response(self, entity, request, response, prefix=''):
+        # type: (RateLimitedObject, HttpRequest, HttpResponse, Text) -> HttpResponse
+        if not (settings.RATE_LIMITING or settings.IP_RATE_LIMITING or
+                settings.EMAIL_RATE_LIMITING):
             return response
 
-        from zerver.lib.rate_limiter import max_api_calls, RateLimitedUser
+        header_prefix = prefix.title().replace('_', '-')
         # Add X-RateLimit-*** headers
-        if hasattr(request, '_ratelimit_applied_limits'):
-            entity = RateLimitedUser(request.user)
-            response['X-RateLimit-Limit'] = str(max_api_calls(entity))
-            if hasattr(request, '_ratelimit_secs_to_freedom'):
-                response['X-RateLimit-Reset'] = str(int(time.time() + request._ratelimit_secs_to_freedom))
-            if hasattr(request, '_ratelimit_remaining'):
-                response['X-RateLimit-Remaining'] = str(request._ratelimit_remaining)
+        attr = '{}_ratelimit_applied_limits'.format(prefix)
+        if hasattr(request, attr):
+            header = 'X{}-RateLimit-Limit'.format(header_prefix)
+            response[header] = str(max_api_calls(entity))
+            attr = '{}_ratelimit_secs_to_freedom'.format(prefix)
+            if hasattr(request, attr):
+                header = 'X{}-RateLimit-Reset'.format(header_prefix)
+                response[header] = str(int(time.time() + getattr(request, attr)))
+            attr = '{}_ratelimit_remaining'.format(prefix)
+            if hasattr(request, attr):
+                header = 'X{}-RateLimit-Remaining'.format(header_prefix)
+                response[header] = str(getattr(request, attr))
         return response
 
     def process_exception(self, request, exception):
         # type: (HttpRequest, Exception) -> Optional[HttpResponse]
         if isinstance(exception, RateLimited):
+            secs_to_freedom = exception.args[0]
```

Comment:
So in this spot we switch from attributes on the request to using data in an exception, but above we're still using attributes on the request.

I think leaving things as attributes on the request is fine. We definitely shouldn't be mixing the two dataflow paths, though, unless there's some clear reason for it -- that kind of duplication just invites bugs by letting things get out of sync.